internal/services/services.go

```diff
@@ -77,5 +89,73 @@ func toggleServices(
 			}
 		}
 	}
+
+	waitGroup.Wait()
 	return nil
 }
+
+func listenToAutoPortForwardingChangesOnRemote(
+	ctx context.Context,
+	serviceName string,
+	w io.Writer,
+	projectDir string,
+	action serviceAction,
+	waitGroup *sync.WaitGroup,
+) error {
+
+	if os.Getenv("DEVBOX_REGION") == "" {
+		return nil
+	}
+	hostname, err := hostname()
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	fmt.Fprintf(w, "Waiting for port forwarding to start/stop for service %q\n", serviceName)
+	waitGroup.Add(1)
+
+	ctx, cancel := context.WithCancel(ctx)
+
+	// Done function is thread safe and will cancel the context and decrement the wait group
+	// if context is already canceled, it will not do anything
+	var m sync.Mutex
+	done := func() {
+		m.Lock()
+		defer m.Unlock()
+		if ctx.Err() == nil {
+			cancel()
+			waitGroup.Done()
+		}
+	}
+
+	// After 5 seconds, if the context has not been canceled, go ahead and cancel it
+	// and also decrement the wait group so that the caller can continue.
+	time.AfterFunc(5*time.Second, func() {
```

Comment:
Can this be done with `context.WithTimeout`?

Reply:
Yes, it would need a timeout and also a cancel.